quasiquotes3/shared/src/main/scala/scala/meta/internal/quasiquotes/ConversionMacros.scala

```
@@ -0,0 +1,87 @@
+}
+
+object ConversionMacros {
+  def liftApplyImpl[I: Type](using Quotes)(outside: Expr[Any]) = new ConversionMacros(using quotes).liftApply[I](outside)
+  def liftUnapplyImpl[I: Type](using Quotes)(outside: Expr[Any]) = new ConversionMacros(using quotes).liftUnapply[I](outside)
+
+  def unliftApplyImpl[O: Type](using Quotes)(inside: Expr[Any]) = new ConversionMacros(using quotes).unliftApply[O](inside)
+  def unliftUnapplyImpl[O: Type](using Quotes)(inside: Expr[Any]) = new ConversionMacros(using quotes).unliftUnapply[O](inside)
+}
+
+class ConversionMacros(using val topLevelQuotes: Quotes) {//extends AstReflection {
+  import topLevelQuotes.reflect._
+  // val XtensionQuasiquoteTerm = "shadow scala.meta quasiquotes"
+  import scala.meta.quasiquotes.{Lift, Unlift}
+
+  private def typeMismatchMessage(found: TypeRepr, req: TypeRepr): String = {
+    val foundReqMessage = found.show + ", required " + req.show
+    "type mismatch when unquoting " + foundReqMessage
+  }
+
+  def liftApply[I: Type](outside: Expr[Any]): Expr[I] = {
+    outside match
+      case '{$outsideExpr: o} => 
+        val outsideTpe = TypeRepr.of[o].widen
+        val insideTpe = TypeRepr.of[I]
+        if (outsideTpe <:< insideTpe) {
+          outside.asExprOf[I]
+        } else {
+          outsideTpe.asType match
+            case '[t] =>
+              val liftableMaybe = Expr.summon[Lift[t, I]]
+              liftableMaybe match
+                case Some(liftable) => '{$liftable.apply(${outsideExpr.asExprOf[t]})}
+                case None => report.errorAndAbort(typeMismatchMessage(outsideTpe, insideTpe))
+        }
+  }
+
+  def liftUnapply[I: Type](outside: Expr[Any]): Expr[Option[I]] = {
+    // NOTE: Here's an interesting idea that I'd like to explore.
+    // How about we allow things like `42 match { case q"$x" => x }`?
+    // For that to work, we just need to wrap the reification result into `Lift.unapply`!
+    ???
```

Let's not leave it at `???` better to throw an exception with an explanation.

---

Same here as below, copied from the counterpart implementation for Scala 2. I can add a legitimate message, but then it would propably be better to also add it into the Scala 2 version of the file